cpp/include/cuml/fil/fnv_hash.h

```diff
@@ -0,0 +1,32 @@
+/* Please do not copyright this code.  This code is in the public domain.
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <limits.h>
+#include <numeric>
+
+// Implements https://tools.ietf.org/html/draft-eastlake-fnv-17.html
+// Algorithm is public domain, non-cryptographic strength and no patents or rights to patent.
+// If input elements are not 8-bit, such a computation does not match
+// the FNV spec.
+template <typename It>
+unsigned long long fowler_noll_vo_fingerprint64(It begin, It end) {
+  static_assert(sizeof *begin == 1,
+                "FNV deals with byte-sized (octet) input arrays only");
+  return std::accumulate(begin, end, 14695981039346656037ull,
+                         [](const unsigned long long& fingerprint, auto x) {
+                           return (fingerprint * 0x100000001b3ull) ^ x;
+                         });
+}
+
+// xor-folded fingerprint64 to ensure first bits are affected by other input bits
+// should give a 1% collision probability within a 10'000 hash set
+template <typename It>
+unsigned long fowler_noll_vo_fingerprint64_32(It begin, It end) {
```

Comment:
`unsigned long` may be 32- or 64-bit, depending on the compiler. If a 32-bit unsigned integer type is intended, it's better to use `uint32_t`.